{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da192129-6fa8-4123-bd7c-39a192ded118",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Libraries\n",
    "\n",
    "import matplotlib\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "import platform\n",
    "import scipy\n",
    "import seaborn\n",
    "import shutil\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea5e4899-dd47-4333-be19-b1ce72ecab81",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ensure source path\n",
    "\n",
    "ROOT = os.getcwd()\n",
    "\n",
    "while not ROOT.endswith('upolanc-thesis') :\n",
    "\tROOT = os.path.abspath(os.path.join(ROOT, os.pardir))\n",
    "\n",
    "\tif len(ROOT) < len('upolanc-thesis') :\n",
    "\t\tif   platform.system() == 'Linux'   : ROOT = '/d/hpc/projects/FRI/up4472/upolanc-thesis'\n",
    "\t\telif platform.system() == 'Windows' : ROOT = 'C:\\\\Developer\\\\Workspace\\\\PyCharm\\\\Projects\\\\upolanc-thesis'\n",
    "\t\telse : raise ValueError()\n",
    "\n",
    "\t\tprint(f'Warning : could not find correct directory, using default : {ROOT}')\n",
    "\t\tprint()\n",
    "\n",
    "\t\tbreak\n",
    "\n",
    "if ROOT not in sys.path :\n",
    "\tsys.path.append(ROOT)\n",
    "\n",
    "os.chdir(ROOT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "047967c3-229a-47ed-94bb-8a1b0e3b5ff3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Code\n",
    "\n",
    "from source.python              import runtime\n",
    "from source.python.cnn          import cnn_plot\n",
    "from source.python.data.feature import feature_processing\n",
    "from source.python.io           import loader\n",
    "\n",
    "runtime.set_numpy_format()\n",
    "runtime.set_pandas_format()\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6038e741-268e-4b3d-854e-eeea80f17ed1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-09-21T11:41:59.518014Z",
     "iopub.status.busy": "2023-09-21T11:41:59.517514Z",
     "iopub.status.idle": "2023-09-21T11:41:59.532026Z",
     "shell.execute_reply": "2023-09-21T11:41:59.531525Z",
     "shell.execute_reply.started": "2023-09-21T11:41:59.518014Z"
    }
   },
   "source": [
    "# 1. Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "212c9e08-3d2b-45d9-9bb4-d1b3ed031793",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for args\n",
    "\n",
    "if __name__ == '__main__' and 'get_ipython' in dir() :\n",
    "\tprint('Running as .ipynb')\n",
    "\tprint()\n",
    "\n",
    "if __name__ == '__main__' and 'get_ipython' not in dir() :\n",
    "\tprint('Running as .py')\n",
    "\tprint()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c30195f2-b0c9-4c0f-bdd2-43bdc45e184e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Setup some directory paths\n",
    "\n",
    "CWD = ROOT\n",
    "OUT = os.path.join(CWD, 'output')\n",
    "RES = os.path.join(CWD, 'resources')\n",
    "\n",
    "OUT_DATA  = os.path.join(OUT, 'nbp17-plots')\n",
    "RES_NBP01 = os.path.join(OUT, 'nbp01-filter')\n",
    "RES_NBP02 = os.path.join(OUT, 'nbp02-anndata')\n",
    "RES_NBP04 = os.path.join(OUT, 'nbp04-feature')\n",
    "\n",
    "shutil.rmtree(OUT_DATA, ignore_errors = True)\n",
    "\n",
    "os.makedirs(OUT_DATA, exist_ok = True)\n",
    "\n",
    "print(f'     Root Directory : {CWD}')\n",
    "print(f'   Output Directory : {OUT_DATA}')\n",
    "print(f' Resource Directory : {RES_NBP01}')\n",
    "print(f' Resource Directory : {RES_NBP02}')\n",
    "print(f' Resource Directory : {RES_NBP04}')\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc7ce989-891a-428d-85a0-99406094bf98",
   "metadata": {},
   "source": [
    "# 2. Transcript RSD Distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cd5224c-b2bf-40f6-b437-74384181e3a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Compute RSD for every filter\n",
    "\n",
    "dataframes = dict()\n",
    "\n",
    "for fid in [1, 2, 3, 4, 5, 6] :\n",
    "\tanndata = loader.load_h5ad(filename = os.path.join(RES_NBP04, 'filter{}'.format(fid), 'arabidopsis-r36.h5ad'))\n",
    "\tsplit   = loader.load_json(filename = os.path.join(RES_NBP01, 'filter{}'.format(fid), 'filter.json'))\n",
    "\n",
    "\tkeep_transcript = split['data']['keep_transcript']\n",
    "\tdrop_transcript = split['data']['drop_transcript']\n",
    "\n",
    "\tresult = dict()\n",
    "\n",
    "\tresult['Data']      = list()\n",
    "\tresult['Group']     = list()\n",
    "\tresult['Target']    = list()\n",
    "\tresult['Median']    = list()\n",
    "\tresult['Mean']      = list()\n",
    "\tresult['Min']       = list()\n",
    "\tresult['Max']       = list()\n",
    "\tresult['Samples']   = list()\n",
    "\tresult['RSD']       = list()\n",
    "\tresult['RSD < 1.0'] = list()\n",
    "\tresult['RSD < 0.5'] = list()\n",
    "\n",
    "\ttarget = 'global-mean'\n",
    "\tmatrix = anndata\n",
    "\n",
    "\tfor selection in ['all', 'keep', 'drop'] :\n",
    "\t\tif   selection == 'all'  : data = matrix\n",
    "\t\telif selection == 'keep' : data = matrix[:, anndata.var.index.isin(keep_transcript)].copy()\n",
    "\t\telif selection == 'drop' : data = matrix[:, anndata.var.index.isin(drop_transcript)].copy()\n",
    "\t\telse : raise ValueError()\n",
    "\n",
    "\t\tdata = data.layers['boxcox1p']\n",
    "\t\tmean = data.mean(axis = 0)\n",
    "\t\tstd  = data.std(axis = 0)\n",
    "\t\trsd  = std / mean\n",
    "\n",
    "\t\tresult['Data'     ].append(selection)\n",
    "\t\tresult['Target'   ].append(target)\n",
    "\t\tresult['Group'    ].append('')\n",
    "\t\tresult['RSD'      ].append(rsd)\n",
    "\t\tresult['Median'   ].append(numpy.median(rsd))\n",
    "\t\tresult['Mean'     ].append(numpy.mean(rsd))\n",
    "\t\tresult['Min'      ].append(numpy.min(rsd))\n",
    "\t\tresult['Max'      ].append(numpy.max(rsd))\n",
    "\t\tresult['Samples'  ].append(numpy.size(data, 0))\n",
    "\t\tresult['RSD < 1.0'].append(numpy.sum(rsd < 1.0) / len(rsd))\n",
    "\t\tresult['RSD < 0.5'].append(numpy.sum(rsd < 0.5) / len(rsd))\n",
    "\n",
    "\ttarget = 'tissue-mean'\n",
    "\tmatrix = anndata\n",
    "\n",
    "\tfor group in anndata.obs['Tissue'].unique() :\n",
    "\t\tmatrix = anndata[anndata.obs['Tissue'] == group, :].copy()\n",
    "\n",
    "\t\tfor selection in ['all', 'keep', 'drop'] :\n",
    "\t\t\tif   selection == 'all'  : data = matrix\n",
    "\t\t\telif selection == 'keep' : data = matrix[:, anndata.var.index.isin(keep_transcript)].copy()\n",
    "\t\t\telif selection == 'drop' : data = matrix[:, anndata.var.index.isin(drop_transcript)].copy()\n",
    "\t\t\telse : raise ValueError()\n",
    "\n",
    "\t\t\tdata = data.layers['boxcox1p']\n",
    "\t\t\tmean = data.mean(axis = 0)\n",
    "\t\t\tstd  = data.std(axis = 0)\n",
    "\t\t\trsd  = std / mean\n",
    "\n",
    "\t\t\tresult['Data'     ].append(selection)\n",
    "\t\t\tresult['Target'   ].append(target)\n",
    "\t\t\tresult['Group'    ].append(group)\n",
    "\t\t\tresult['RSD'      ].append(rsd)\n",
    "\t\t\tresult['Median'   ].append(numpy.median(rsd))\n",
    "\t\t\tresult['Mean'     ].append(numpy.mean(rsd))\n",
    "\t\t\tresult['Min'      ].append(numpy.min(rsd))\n",
    "\t\t\tresult['Max'      ].append(numpy.max(rsd))\n",
    "\t\t\tresult['Samples'  ].append(numpy.size(data, 0))\n",
    "\t\t\tresult['RSD < 1.0'].append(numpy.sum(rsd < 1.0) / len(rsd))\n",
    "\t\t\tresult['RSD < 0.5'].append(numpy.sum(rsd < 0.5) / len(rsd))\n",
    "\n",
    "\ttarget = 'group-mean'\n",
    "\tmatrix = anndata\n",
    "\n",
    "\tfor group in anndata.obs['Group'].unique() :\n",
    "\t\tmatrix = anndata[anndata.obs['Group'] == group, :].copy()\n",
    "\n",
    "\t\tfor selection in ['all', 'keep', 'drop'] :\n",
    "\t\t\tif   selection == 'all'  : data = matrix\n",
    "\t\t\telif selection == 'keep' : data = matrix[:, anndata.var.index.isin(keep_transcript)].copy()\n",
    "\t\t\telif selection == 'drop' : data = matrix[:, anndata.var.index.isin(drop_transcript)].copy()\n",
    "\t\t\telse : raise ValueError()\n",
    "\n",
    "\t\t\tdata = data.layers['boxcox1p']\n",
    "\t\t\tmean = data.mean(axis = 0)\n",
    "\t\t\tstd  = data.std(axis = 0)\n",
    "\t\t\trsd  = std / mean\n",
    "\n",
    "\t\t\tresult['Data'     ].append(selection)\n",
    "\t\t\tresult['Target'   ].append(target)\n",
    "\t\t\tresult['Group'    ].append(group)\n",
    "\t\t\tresult['RSD'      ].append(rsd)\n",
    "\t\t\tresult['Median'   ].append(numpy.median(rsd))\n",
    "\t\t\tresult['Mean'     ].append(numpy.mean(rsd))\n",
    "\t\t\tresult['Min'      ].append(numpy.min(rsd))\n",
    "\t\t\tresult['Max'      ].append(numpy.max(rsd))\n",
    "\t\t\tresult['Samples'  ].append(numpy.size(data, 0))\n",
    "\t\t\tresult['RSD < 1.0'].append(numpy.sum(rsd < 1.0) / len(rsd))\n",
    "\t\t\tresult['RSD < 0.5'].append(numpy.sum(rsd < 0.5) / len(rsd))\n",
    "\n",
    "\tdataframes[fid] = pandas.DataFrame.from_dict(result)\n",
    "\tdataframes[fid].to_csv(os.path.join(OUT_DATA, 'table-rsd-f{}.csv'.format(fid)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43035f7e-7fb6-4f7d-a583-16f627f79621",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Display RSD\n",
    "\n",
    "search_tissue = [\n",
    "\t('tissue-mean', 'seedling'),\n",
    "\t('tissue-mean', 'seed'),\n",
    "\t('tissue-mean', 'leaf'),\n",
    "\t('tissue-mean', 'root'),\n",
    "\t('tissue-mean', 'flower')\n",
    "]\n",
    "\n",
    "search_group = [\n",
    "\t('group-mean', 'young_seedling'),\n",
    "\t('group-mean', 'mature_seed'),\n",
    "\t('group-mean', 'mature_leaf'),\n",
    "\t('group-mean', 'mature_root'),\n",
    "\t('group-mean', 'mature_flower')\n",
    "]\n",
    "\n",
    "search = [('global-mean', None)]\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 30})\n",
    "seaborn.set_theme(font_scale = 2.5)\n",
    "\n",
    "search.extend(search_tissue)\n",
    "search.extend(search_group)\n",
    "\n",
    "for fid in [1, 2, 3, 4, 5, 6] :\n",
    "\tfig, ax = matplotlib.pyplot.subplots(figsize = (16, 10))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tcolor = {\n",
    "\t\t'global-mean' : 'r',\n",
    "\t\t'tissue-mean' : 'g',\n",
    "\t\t'group-mean'  : 'b'\n",
    "\t}\n",
    "\n",
    "\tfor target, group in search :\n",
    "\t\tx = dataframes[fid]\n",
    "\t\tx = x[x['Data'] == 'all']\n",
    "\n",
    "\t\tif target is not None : x = x[x['Target'] == target]\n",
    "\t\tif group  is not None : x = x[x['Group']  == group]\n",
    "\n",
    "\t\tx = x['RSD'].iloc[0]\n",
    "\n",
    "\t\trsd_x = list()\n",
    "\t\trsd_y = list()\n",
    "\n",
    "\t\tfor t in numpy.arange(0.00, 2.00, 0.05) :\n",
    "\t\t\trsd_y.append(t)\n",
    "\t\t\trsd_x.append(numpy.sum(x < t) / len(x) * 100.0)\n",
    "\n",
    "\t\tlabel = ''\n",
    "\n",
    "\t\tif target is not None : label = target\n",
    "\t\tif group  is not None : label = label + '-' + group\n",
    "\n",
    "\t\tseaborn.lineplot(\n",
    "\t\t\tx         = rsd_y,\n",
    "\t\t\ty         = rsd_x,\n",
    "\t\t\tlabel     = label,\n",
    "\t\t\tax        = ax,\n",
    "\t\t\talpha     = 0.9,\n",
    "\t\t\tcolor     = color[target],\n",
    "\t\t\tlinewidth = 4\n",
    "\t\t)\n",
    "\n",
    "\tmatplotlib.pyplot.legend(\n",
    "\t\tloc  = 'lower right',\n",
    "\t\tprop = None\n",
    "\t)\n",
    "\n",
    "\tax.set_xlabel('RSD')\n",
    "\tax.set_ylabel('Percentage')\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, 'graph-rsd-f{}.png'.format(fid)),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c78903d1-d556-4e35-a06b-e7e09a4630d8",
   "metadata": {},
   "source": [
    "# 3. Variance vs. TPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c663eb8-636b-455d-be03-5597416e4690",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute variance\n",
    "\n",
    "anndata = loader.load_h5ad(filename = os.path.join(RES_NBP04, 'filter2', 'arabidopsis-r36.h5ad'))\n",
    "filters = loader.load_json(filename = os.path.join(RES_NBP01, 'filter2', 'filter.json'))\n",
    "anndata = anndata[:, anndata.var.index.isin(filters['data']['keep_transcript'])]\n",
    "\n",
    "dictionary = dict()\n",
    "supergroup = 'Tissue'\n",
    "layer      = 'boxcox1p'\n",
    "\n",
    "for group in anndata.obs[supergroup].unique() :\n",
    "\tsubdata = anndata[anndata.obs[supergroup].isin([group]), :]\n",
    "\n",
    "\ttranscript = subdata.var.index.tolist()\n",
    "\tvariance   = subdata.layers[layer].var(axis = 1).tolist()\n",
    "\tstdev      = subdata.layers[layer].std(axis = 1).tolist()\n",
    "\tdictionary.setdefault(group, dict())\n",
    "\n",
    "\tfor k, v, s in zip(transcript, variance, stdev) :\n",
    "\t\tdictionary[group].setdefault(k, dict()).setdefault('variance', v)\n",
    "\t\tdictionary[group].setdefault(k, dict()).setdefault('stdev', s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fc97fef-b576-47d6-8ddf-683a3128e71f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot scatterplot variance\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\reports\\regression-cnn'\n",
    "models_path = os.path.join(folder_path, 'washburn-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "report_path = os.path.join(models_path, 'report_keep.json')\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 36})\n",
    "seaborn.set_theme(font_scale = 3)\n",
    "\n",
    "for xtarget in ['ypred', 'ytrue'] :\n",
    "\treport = loader.load_json(report_path)\n",
    "\treport = report['eval']\n",
    "\n",
    "\tscatterdict = {\n",
    "\t\t'ypred' : list(),\n",
    "\t\t'ytrue' : list(),\n",
    "\t\t'value' : list(),\n",
    "\t\t'group' : list()\n",
    "\t}\n",
    "\n",
    "\tfor key, ypred, ytrue in zip(report['keys'], report['ypred'], report['ytrue']) :\n",
    "\t\tif isinstance(ypred, list) : ypred = ypred[0]\n",
    "\t\tif isinstance(ytrue, list) : ytrue = ytrue[0]\n",
    "\n",
    "\t\tgroup      = key.split('?')[0]\n",
    "\t\ttranscript = key.split('?')[1]\n",
    "\n",
    "\t\tif not group      in dictionary.keys()        : continue\n",
    "\t\tif not transcript in dictionary[group].keys() : continue\n",
    "\n",
    "\t\tscatterdict['ytrue'].append(ytrue)\n",
    "\t\tscatterdict['ypred'].append(ypred)\n",
    "\t\tscatterdict['value'].append(dictionary[group][transcript]['variance'])\n",
    "\t\tscatterdict['group'].append(group)\n",
    "\n",
    "\tscatterdf = pandas.DataFrame.from_dict(scatterdict)\n",
    "\tscatterdf = scatterdf.rename(columns = {\n",
    "\t\txtarget : 'TPM',\n",
    "\t\t'value' : 'Variance',\n",
    "\t\t'group' : 'Tissue'\n",
    "\t})\n",
    "\n",
    "\tfig, ax = matplotlib.pyplot.subplots(figsize = (14, 10))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tseaborn.scatterplot(\n",
    "\t\tdata  = scatterdf,\n",
    "\t\tx     = 'TPM',\n",
    "\t\ty     = 'Variance',\n",
    "\t\thue   = 'Tissue',\n",
    "\t\tax    = ax,\n",
    "\t\ts     = 300,\n",
    "\t\talpha = 0.9\n",
    "\t)\n",
    "\n",
    "\tax.set_xlabel('TPM')\n",
    "\tax.set_ylabel('Variance')\n",
    "\tax.legend(markerscale = 2)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, 'graph-{}-scatterplot.png'.format(xtarget)),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23ef1c51-d9be-4ec4-9c93-c106eda5d227",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot lmplot variance\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\reports\\regression-cnn'\n",
    "models_path = os.path.join(folder_path, 'washburn-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "report_path = os.path.join(models_path, 'report_keep.json')\n",
    "\n",
    "for xtarget in ['ypred', 'ytrue'] :\n",
    "\treport = loader.load_json(report_path)\n",
    "\treport = report['eval']\n",
    "\n",
    "\tscatterdict = {\n",
    "\t\t'ypred' : list(),\n",
    "\t\t'ytrue' : list(),\n",
    "\t\t'value' : list(),\n",
    "\t\t'group' : list()\n",
    "\t}\n",
    "\n",
    "\tfor key, ypred, ytrue in zip(report['keys'], report['ypred'], report['ytrue']) :\n",
    "\t\tif isinstance(ypred, list) : ypred = ypred[0]\n",
    "\t\tif isinstance(ytrue, list) : ytrue = ytrue[0]\n",
    "\n",
    "\t\tgroup      = key.split('?')[0]\n",
    "\t\ttranscript = key.split('?')[1]\n",
    "\n",
    "\t\tif not group      in dictionary.keys()        : continue\n",
    "\t\tif not transcript in dictionary[group].keys() : continue\n",
    "\n",
    "\t\tscatterdict['ytrue'].append(ytrue)\n",
    "\t\tscatterdict['ypred'].append(ypred)\n",
    "\t\tscatterdict['value'].append(dictionary[group][transcript]['variance'])\n",
    "\t\tscatterdict['group'].append(group)\n",
    "\n",
    "\tscatterdf = pandas.DataFrame.from_dict(scatterdict)\n",
    "\tscatterdf = scatterdf.rename(columns = {\n",
    "\t\txtarget : 'TPM',\n",
    "\t\t'value' : 'Variance',\n",
    "\t\t'group' : 'Tissue'\n",
    "\t})\n",
    "\n",
    "\tseaborn.lmplot(\n",
    "\t\tdata  = scatterdf,\n",
    "\t\tx     = 'TPM',\n",
    "\t\ty     = 'Variance',\n",
    "\t\thue   = 'Tissue',\n",
    "\t\theight = 10,\n",
    "\t\taspect = 1.6\n",
    "\t)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, 'graph-{}-lmplot.png'.format(xtarget)),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c0050da-7c4b-43ac-a0df-833371265ead",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot boxplot variance\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\reports\\regression-cnn'\n",
    "models_path = os.path.join(folder_path, 'washburn-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "report_path = os.path.join(models_path, 'report_keep.json')\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 36})\n",
    "seaborn.set_theme(font_scale = 3)\n",
    "\n",
    "for xtarget in ['ypred', 'ytrue'] :\n",
    "\treport = loader.load_json(report_path)\n",
    "\treport = report['eval']\n",
    "\n",
    "\tscatterdict = {\n",
    "\t\t'ypred' : list(),\n",
    "\t\t'ytrue' : list(),\n",
    "\t\t'value' : list(),\n",
    "\t\t'group' : list()\n",
    "\t}\n",
    "\n",
    "\tfor key, ypred, ytrue in zip(report['keys'], report['ypred'], report['ytrue']) :\n",
    "\t\tif isinstance(ypred, list) : ypred = ypred[0]\n",
    "\t\tif isinstance(ytrue, list) : ytrue = ytrue[0]\n",
    "\n",
    "\t\tgroup      = key.split('?')[0]\n",
    "\t\ttranscript = key.split('?')[1]\n",
    "\n",
    "\t\tif not group      in dictionary.keys()        : continue\n",
    "\t\tif not transcript in dictionary[group].keys() : continue\n",
    "\n",
    "\t\tscatterdict['ytrue'].append(ytrue)\n",
    "\t\tscatterdict['ypred'].append(ypred)\n",
    "\t\tscatterdict['value'].append(dictionary[group][transcript]['variance'])\n",
    "\t\tscatterdict['group'].append(group)\n",
    "\n",
    "\tscatterdf = pandas.DataFrame.from_dict(scatterdict)\n",
    "\tscatterdf = scatterdf.rename(columns = {\n",
    "\t\txtarget : 'TPM',\n",
    "\t\t'value' : 'Variance',\n",
    "\t\t'group' : 'Tissue'\n",
    "\t})\n",
    "\n",
    "\tfig, ax = matplotlib.pyplot.subplots(figsize = (7, 12))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tseaborn.boxplot(\n",
    "\t\tdata  = scatterdf,\n",
    "\t\tx     = 'Variance',\n",
    "\t\ty     = 'Tissue',\n",
    "\t\tax    = ax\n",
    "\t)\n",
    "\n",
    "\tax.set_xlabel('Variance')\n",
    "\tax.set_ylabel('Tissue')\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, 'graph-{}-boxplot.png'.format(xtarget)),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa6d89dc-196c-4a2b-a365-c0de524d2e72",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot scatterplot and boxplot variance\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\reports\\regression-cnn'\n",
    "models_path = os.path.join(folder_path, 'washburn-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "report_path = os.path.join(models_path, 'report_keep.json')\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 36})\n",
    "seaborn.set_theme(font_scale = 3)\n",
    "\n",
    "for xtarget in ['ypred', 'ytrue'] :\n",
    "\treport = loader.load_json(report_path)\n",
    "\treport = report['eval']\n",
    "\n",
    "\tscatterdict = {\n",
    "\t\t'ypred' : list(),\n",
    "\t\t'ytrue' : list(),\n",
    "\t\t'value' : list(),\n",
    "\t\t'group' : list()\n",
    "\t}\n",
    "\n",
    "\tfor key, ypred, ytrue in zip(report['keys'], report['ypred'], report['ytrue']) :\n",
    "\t\tif isinstance(ypred, list) : ypred = ypred[0]\n",
    "\t\tif isinstance(ytrue, list) : ytrue = ytrue[0]\n",
    "\n",
    "\t\tgroup      = key.split('?')[0]\n",
    "\t\ttranscript = key.split('?')[1]\n",
    "\n",
    "\t\tif not group      in dictionary.keys()        : continue\n",
    "\t\tif not transcript in dictionary[group].keys() : continue\n",
    "\n",
    "\t\tscatterdict['ytrue'].append(ytrue)\n",
    "\t\tscatterdict['ypred'].append(ypred)\n",
    "\t\tscatterdict['value'].append(dictionary[group][transcript]['variance'])\n",
    "\t\tscatterdict['group'].append(group)\n",
    "\n",
    "\tscatterdf = pandas.DataFrame.from_dict(scatterdict)\n",
    "\tscatterdf = scatterdf.rename(columns = {\n",
    "\t\txtarget : 'TPM',\n",
    "\t\t'value' : 'Variance',\n",
    "\t\t'group' : 'Tissue'\n",
    "\t})\n",
    "\n",
    "\tfig, ax = matplotlib.pyplot.subplots(\n",
    "\t\tnrows       = 1,\n",
    "\t\tncols       = 2,\n",
    "\t\tsharey      = True,\n",
    "\t\tfigsize     = (28, 12),\n",
    "\t\tgridspec_kw = {\n",
    "\t\t\t'width_ratios' : [60, 40]\n",
    "\t\t}\n",
    "\t)\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tax1 = ax[0]\n",
    "\tax2 = ax[1]\n",
    "\n",
    "\tseaborn.scatterplot(\n",
    "\t\tdata  = scatterdf,\n",
    "\t\tx     = 'TPM',\n",
    "\t\ty     = 'Variance',\n",
    "\t\thue   = 'Tissue',\n",
    "\t\tax    = ax1,\n",
    "\t\ts     = 300,\n",
    "\t\talpha = 0.9\n",
    "\t)\n",
    "\n",
    "\tax1.set_xlabel('TPM')\n",
    "\tax1.set_ylabel('Variance')\n",
    "\tax1.legend(markerscale = 2)\n",
    "\n",
    "\tseaborn.boxplot(\n",
    "\t\tdata  = scatterdf,\n",
    "\t\ty     = 'Variance',\n",
    "\t\tx     = 'Tissue',\n",
    "\t\tax    = ax2\n",
    "\t)\n",
    "\n",
    "\tax2.set_xlabel('Tissue')\n",
    "\tax2.set_ylabel(None)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, 'graph-{}-multiplot.png'.format(xtarget)),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "660f04e8-bd45-465b-897b-dbf18d649530",
   "metadata": {},
   "source": [
    "# 4. Prediction Error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09905298-4133-41bc-9c53-79f18fbd9997",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot prediction error\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\reports\\regression-cnn'\n",
    "wmodels_path = os.path.join(folder_path,  'washburn-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "zmodels_path = os.path.join(folder_path,  'zrimec-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "wreport_path = os.path.join(wmodels_path, 'report_keep.json')\n",
    "zreport_path = os.path.join(zmodels_path, 'report_keep.json')\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 48})\n",
    "seaborn.set_theme(font_scale = 4)\n",
    "\n",
    "for name, report_path in [('washburn', wreport_path), ('zrimec', zreport_path)] :\n",
    "\treport = loader.load_json(report_path)\n",
    "\treport = report['eval']\n",
    "\n",
    "\tkeys = set([x.split('?')[0] for x in report['keys']])\n",
    "\tvals = dict()\n",
    "\n",
    "\tfor key in keys :\n",
    "\t\tvals[key] = list()\n",
    "\n",
    "\tfor key, ypred, ytrue in zip(report['keys'], report['ypred'], report['ytrue']) :\n",
    "\t\tif isinstance(ypred, list) : ypred = ypred[0]\n",
    "\t\tif isinstance(ytrue, list) : ytrue = ytrue[0]\n",
    "\n",
    "\t\tkey = key.split('?')[0]\n",
    "\t\tval = ypred - ytrue\n",
    "\n",
    "\t\tvals[key].append(val)\n",
    "\n",
    "\tn, nrows, ncols = cnn_plot.compute_gridsize(\n",
    "\t\tn = len(keys)\n",
    "\t)\n",
    "\n",
    "\tif nrows < ncols :\n",
    "\t\tnrows, ncols = ncols, nrows\n",
    "\n",
    "\tfig, ax = matplotlib.pyplot.subplots(nrows, ncols, sharex = True, sharey = True, figsize = (10 * ncols, 10 * nrows))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tminval = 0\n",
    "\tmaxval = 0\n",
    "\n",
    "\tfor index, key in enumerate(keys) :\n",
    "\t\tminval = min(numpy.min(vals[key]), minval)\n",
    "\t\tmaxval = max(numpy.max(vals[key]), maxval)\n",
    "\n",
    "\tfor index, key in enumerate(sorted(keys)) :\n",
    "\t\tr = index // ncols\n",
    "\t\tc = index  % ncols\n",
    "\n",
    "\t\taxis = ax[r, c]\n",
    "\n",
    "\t\tseaborn.histplot(\n",
    "\t\t\tx     = vals[key],\n",
    "\t\t\tax    = axis,\n",
    "\t\t\talpha = 0.9\n",
    "\t\t)\n",
    "\n",
    "\t\taxis.axvline(x = 0, color = 'r', linewidth = 4)\n",
    "\n",
    "\t\tgmin = minval - 0.1\n",
    "\t\tgmax = 290\n",
    "\n",
    "\t\taxis.text(gmin, 1.00 * gmax, r'$\\bar{x}$')\n",
    "\t\taxis.text(gmin, 0.89 * gmax, r'$\\tilde{x}$')\n",
    "\t\n",
    "\t\taxis.text(gmin + 0.38, 1.00 * gmax, '=')\n",
    "\t\taxis.text(gmin + 0.38, 0.89 * gmax, '=')\n",
    "\n",
    "\t\taxis.text(gmin + 0.65, 1.00 * gmax, '{: .3f}'.format(numpy.mean(vals[key])))\n",
    "\t\taxis.text(gmin + 0.65, 0.89 * gmax, '{: .3f}'.format(numpy.median(vals[key])))\n",
    "\n",
    "\t\taxis.set_title(key.title())\n",
    "\t\taxis.set_xlabel('Error')\n",
    "\t\taxis.set_ylabel('Count')\n",
    "\n",
    "\tfor index in range(n, nrows * ncols) :\n",
    "\t\tif nrows == 1 or ncols == 1 : axis = ax[index]\n",
    "\t\telse                        : axis = ax[index // ncols, index % ncols]\n",
    "\n",
    "\t\taxis.axis('off')\n",
    "\n",
    "\tmatplotlib.pyplot.subplots_adjust(\n",
    "\t\tleft   = None,\n",
    "\t\tbottom = None,\n",
    "\t\tright  = None,\n",
    "\t\ttop    = None,\n",
    "\t\twspace = 0.05,\n",
    "\t\thspace = 0.10\n",
    "\t)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, 'graph-{}-prediction-error.png'.format(name)),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22c90a71-4b89-44b7-a36d-bd6794cec875",
   "metadata": {},
   "source": [
    "# 5. Actual vs. Predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a6e83b2-e0b7-4407-9440-0a9b43a42e4b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot actual vs. predicted\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\reports\\regression-cnn'\n",
    "wmodels_path = os.path.join(folder_path,  'washburn-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "zmodels_path = os.path.join(folder_path,  'zrimec-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "wreport_path = os.path.join(wmodels_path, 'report_keep.json')\n",
    "zreport_path = os.path.join(zmodels_path, 'report_keep.json')\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 48})\n",
    "seaborn.set_theme(font_scale = 4)\n",
    "\n",
    "for name, report_path in [('washburn', wreport_path), ('zrimec', zreport_path)] :\n",
    "\treport = loader.load_json(report_path)\n",
    "\treport = report['eval']\n",
    "\n",
    "\tkeys   = set([x.split('?')[0] for x in report['keys']])\n",
    "\typreds = dict()\n",
    "\tytrues = dict()\n",
    "\n",
    "\tfor key in keys :\n",
    "\t\typreds[key] = list()\n",
    "\t\tytrues[key] = list()\n",
    "\n",
    "\tfor key, ypred, ytrue in zip(report['keys'], report['ypred'], report['ytrue']) :\n",
    "\t\tif isinstance(ypred, list) : ypred = ypred[0]\n",
    "\t\tif isinstance(ytrue, list) : ytrue = ytrue[0]\n",
    "\n",
    "\t\tkey = key.split('?')[0]\n",
    "\n",
    "\t\typreds[key].append(ypred)\n",
    "\t\tytrues[key].append(ytrue)\n",
    "\n",
    "\tn, nrows, ncols = cnn_plot.compute_gridsize(\n",
    "\t\tn = len(keys)\n",
    "\t)\n",
    "\n",
    "\tif nrows < ncols :\n",
    "\t\tnrows, ncols = ncols, nrows\n",
    "\n",
    "\tfig, ax = matplotlib.pyplot.subplots(nrows, ncols, sharex = True, sharey = True, figsize = (10 * ncols, 10 * nrows))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tfor index, key in enumerate(sorted(keys)) :\n",
    "\t\tr = index // ncols\n",
    "\t\tc = index  % ncols\n",
    "\n",
    "\t\taxis = ax[r, c]\n",
    "\n",
    "\t\tx = numpy.array(ypreds[key])\n",
    "\t\ty = numpy.array(ytrues[key])\n",
    "\n",
    "\t\tseaborn.scatterplot(\n",
    "\t\t\tx     = x,\n",
    "\t\t\ty     = y,\n",
    "\t\t\tax    = axis,\n",
    "\t\t\talpha = 0.9\n",
    "\t\t)\n",
    "\n",
    "\t\tres = scipy.stats.linregress(x, y)\n",
    "\n",
    "\t\taxis.plot(x, res.intercept + res.slope * x,\n",
    "\t\t\tcolor     = 'r',\n",
    "\t\t\tlinewidth = 4\n",
    "\t\t)\n",
    "\n",
    "\t\txmin, xmax = axis.get_xlim()\n",
    "\t\tymin, ymax = axis.get_ylim()\n",
    "\n",
    "\t\tgmin = min(xmin, ymin)\n",
    "\t\tgmax = max(xmax, ymax)\n",
    "\n",
    "\t\taxis.set_title(key.title())\n",
    "\t\taxis.set_xlim([gmin, gmax])\n",
    "\t\taxis.set_ylim([gmin, gmax])\n",
    "\t\taxis.set_aspect('equal')\n",
    "\n",
    "\t\toffset = 0.05\n",
    "\n",
    "\t\taxis.text(0.0 * gmax - offset, 0.92 * gmax, 'k')\n",
    "\t\taxis.text(0.0 * gmax - offset, 0.84 * gmax, 'r')\n",
    "\n",
    "\t\taxis.text(0.05 * gmax - offset, 0.92 * gmax, '=')\n",
    "\t\taxis.text(0.05 * gmax - offset, 0.84 * gmax, '=')\n",
    "\n",
    "\t\taxis.text(0.10 * gmax - offset, 0.92 * gmax, '{: .3f}'.format(res.slope))\n",
    "\t\taxis.text(0.10 * gmax - offset, 0.84 * gmax, '{: .3f}'.format(res.rvalue))\n",
    "\n",
    "\t\taxis.set_xlabel('Predicted')\n",
    "\t\taxis.set_ylabel('Actual')\n",
    "\n",
    "\tfor index in range(n, nrows * ncols) :\n",
    "\t\tif nrows == 1 or ncols == 1 : axis = ax[index]\n",
    "\t\telse                        : axis = ax[index // ncols, index % ncols]\n",
    "\n",
    "\t\taxis.axis('off')\n",
    "\n",
    "\tmatplotlib.pyplot.subplots_adjust(\n",
    "\t\tleft   = None,\n",
    "\t\tbottom = None,\n",
    "\t\tright  = None,\n",
    "\t\ttop    = None,\n",
    "\t\twspace = 0.05,\n",
    "\t\thspace = 0.10\n",
    "\t)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, 'graph-{}-prediction-linefit.png'.format(name)),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac0b60e8-d6fa-4e6b-8820-ec2ae6466d50",
   "metadata": {},
   "source": [
    "# 6. Error vs. TPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b266d565-46eb-4979-9642-9d91484b32cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\reports\\regression-cnn'\n",
    "wmodels_path = os.path.join(folder_path,  'washburn-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "zmodels_path = os.path.join(folder_path,  'zrimec-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "wreport_path = os.path.join(wmodels_path, 'report_keep.json')\n",
    "zreport_path = os.path.join(zmodels_path, 'report_keep.json')\n",
    "\n",
    "data = {\n",
    "\t'washburn' : None,\n",
    "\t'zrimec'   : None,\n",
    "}\n",
    "\n",
    "thresholds_boxcox = numpy.arange(0, 5, 0.25)\n",
    "thresholds_tpm    = [0, 10, 50, 100, 250, 500, 1_000, 5_000, 10_000, 50_000, 250_000, 1_000_000]\n",
    "\n",
    "to_tpm     = lambda x : feature_processing.boxcox1p_inv(x, -0.1455)\n",
    "to_boxcox  = lambda x : x\n",
    "\n",
    "to_absolute_error = lambda ytrue, ypred : abs(ytrue - ypred)\n",
    "to_squared_error  = lambda ytrue, ypred : (ytrue - ypred) ** 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67819203-e25e-4abd-909a-1b86896de3b8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot error vs. tpm\n",
    "\n",
    "combinations = [\n",
    "\t(to_squared_error,  'squared-error'),\n",
    "\t(to_absolute_error, 'absolute-error')\n",
    "]\n",
    "\n",
    "for to_metric, metric_name in combinations :\n",
    "\tdata = {\n",
    "\t\t'washburn' : None,\n",
    "\t\t'zrimec'   : None,\n",
    "\t}\n",
    "\n",
    "\tfor model_name, report_path in [('washburn', wreport_path), ('zrimec', zreport_path)] :\n",
    "\t\treport = loader.load_json(report_path)\n",
    "\t\treport = report['eval']\n",
    "\n",
    "\t\ttemp = {\n",
    "\t\t\t'y' : list(),\n",
    "\t\t\t'x' : list()\n",
    "\t\t}\n",
    "\n",
    "\t\tfor ytrue, ypred in zip(report['ytrue'], report['ypred']) :\n",
    "\t\t\tytrue = ytrue[0]\n",
    "\t\t\typred = ypred[0]\n",
    "\n",
    "\t\t\tytrue = to_boxcox(ytrue)\n",
    "\t\t\typred = to_boxcox(ypred)\n",
    "\t\t\tvalue = to_metric(ytrue, ypred)\n",
    "\n",
    "\t\t\tfor i, threshold in enumerate(thresholds_boxcox) :\n",
    "\t\t\t\tif ytrue <= threshold :\n",
    "\t\t\t\t\ttemp['y'].append(value)\n",
    "\t\t\t\t\ttemp['x'].append(threshold)\n",
    "\n",
    "\t\t\t\t\tbreak\n",
    "\n",
    "\t\tdata[model_name] = temp\n",
    "\n",
    "\tfig, ax = matplotlib.pyplot.subplots(figsize = (16, 10))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = pandas.DataFrame.from_dict(data['washburn']),\n",
    "\t\tx         = 'x',\n",
    "\t\ty         = 'y',\n",
    "\t\tax        = ax,\n",
    "\t\tlinewidth = 4,\n",
    "\t\tlabel     = 'Washburn'\n",
    "\t)\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = pandas.DataFrame.from_dict(data['zrimec']),\n",
    "\t\tx         = 'x',\n",
    "\t\ty         = 'y',\n",
    "\t\tax        = ax,\n",
    "\t\tlinewidth = 4,\n",
    "\t\tlabel     = 'Zrimec'\n",
    "\t)\n",
    "\n",
    "\tax.set_xlabel('TPM')\n",
    "\tax.set_ylabel(metric_name.replace('-', ' ').title())\n",
    "\n",
    "\tmatplotlib.pyplot.legend(\n",
    "\t\tloc = 'upper left'\n",
    "\t)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, 'graph-{}.png'.format(metric_name)),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877ed922-9259-411b-bce3-1aa32a5d0269",
   "metadata": {},
   "source": [
    "# 7. Relevance of Window Occlusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2948bb7c-bb28-42ce-a4ae-67e241fb6a46",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load original occlusion relevance data\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\reports\\regression-cnn'\n",
    "wocclusion_path = os.path.join(folder_path,  'washburn-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "zocclusion_path = os.path.join(folder_path,  'zrimec-0-tf2150-f2-0250-77-tissue-mean-explode')\n",
    "\n",
    "OCCLUSION_SIZE   = 10\n",
    "OCCLUSION_STRIDE = 1\n",
    "OCCLUSION_JSON   = 'table-occlusion-w{}-s{}.json'.format(OCCLUSION_SIZE, OCCLUSION_STRIDE)\n",
    "OCCLUSION_CSV    = 'table-occlusion-w{}-s{}.csv'.format(OCCLUSION_SIZE, OCCLUSION_STRIDE)\n",
    "RELEVANCE_TYPE   = 'R2'\n",
    "\n",
    "wocclusion_json = loader.load_json(os.path.join(wocclusion_path, 'occlusion', OCCLUSION_JSON))\n",
    "zocclusion_json = loader.load_json(os.path.join(zocclusion_path, 'occlusion', OCCLUSION_JSON))\n",
    "wocclusion_csv  = loader.load_csv(os.path.join(wocclusion_path, 'occlusion', OCCLUSION_CSV))\n",
    "zocclusion_csv  = loader.load_csv(os.path.join(zocclusion_path, 'occlusion', OCCLUSION_CSV))\n",
    "\n",
    "wocclusion_json = pandas.DataFrame.from_dict(wocclusion_json)\n",
    "wocclusion_json = wocclusion_json.drop(columns = ['mae-relevance', 'mse-relevance', 'r2-relevance', 'start', 'end'])\n",
    "wocclusion_json = wocclusion_json.rename(columns = {'mae' : 'MAE', 'mse' : 'MSE', 'r2' : 'R2', 'mid' : 'Midpoint'})\n",
    "\n",
    "zocclusion_json = pandas.DataFrame.from_dict(zocclusion_json)\n",
    "zocclusion_json = zocclusion_json.drop(columns = ['mae-relevance', 'mse-relevance', 'r2-relevance', 'start', 'end'])\n",
    "zocclusion_json = zocclusion_json.rename(columns = {'mae' : 'MAE', 'mse' : 'MSE', 'r2' : 'R2', 'mid' : 'Midpoint'})\n",
    "\n",
    "WASHBURN_BASELINE = dict()\n",
    "ZRIMEC_BASELINE   = dict()\n",
    "\n",
    "for metric in ['R2', 'MAE', 'MSE'] :\n",
    "\twr = wocclusion_csv[metric]\n",
    "\tzr = zocclusion_csv[metric]\n",
    "\n",
    "\two = wocclusion_json[metric]\n",
    "\tzo = zocclusion_json[metric]\n",
    "\n",
    "\tZRIMEC_BASELINE[metric]   = (zo / (1 - zr)).to_numpy().mean()\n",
    "\tWASHBURN_BASELINE[metric] = (wo / (1 - wr)).to_numpy().mean()\n",
    "\n",
    "\tprint('Baseline {:3s} for Washburn : {:.5f}'.format(metric, WASHBURN_BASELINE[metric]))\n",
    "\tprint('Baseline {:3s} for Zrimec   : {:.5f}'.format(metric, ZRIMEC_BASELINE[metric]))\n",
    "\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "504e6976-f9c9-4933-be06-9de9cbb345dc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Increase stride artifically\n",
    "\n",
    "OCCLUSION_STRIDE = 10\n",
    "\n",
    "wocclusion_json = wocclusion_json[wocclusion_json['Midpoint'] % OCCLUSION_STRIDE == 0]\n",
    "zocclusion_json = zocclusion_json[zocclusion_json['Midpoint'] % OCCLUSION_STRIDE == 0]\n",
    "\n",
    "wocclusion_csv = wocclusion_csv[wocclusion_csv['Midpoint'] % OCCLUSION_STRIDE == 0]\n",
    "zocclusion_csv = zocclusion_csv[zocclusion_csv['Midpoint'] % OCCLUSION_STRIDE == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce8f469c-7bb0-4931-9915-a6badbc0e8c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot relevance for entire sequence\n",
    "\n",
    "LINEWIDTH = 2\n",
    "ALPHA     = 0.8\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 36})\n",
    "seaborn.set_theme(font_scale = 3)\n",
    "\n",
    "for vline in [True, False] :\n",
    "\tfig, ax = matplotlib.pyplot.subplots(figsize = (16, 8))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = wocclusion_csv,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Washburn',\n",
    "\t\tax        = ax,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = zocclusion_csv,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Zrimec',\n",
    "\t\tax        = ax,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tif vline : \n",
    "\t\tax.axvline(1000, ymin = 0, ymax = 1, color = 'k')\n",
    "\t\tax.axvline(1300, ymin = 0, ymax = 1, color = 'k')\n",
    "\t\tax.axvline(1650, ymin = 0, ymax = 1, color = 'k')\n",
    "\n",
    "\tax.set_xlabel('Position')\n",
    "\tax.set_ylabel('Relevance')\n",
    "\n",
    "\tax.legend(loc = 'upper left')\n",
    "\n",
    "\tif vline : path = 'graph-occlusion-relevance-w{}-s{}-full-with-vline.png'.format(OCCLUSION_SIZE, OCCLUSION_STRIDE)\n",
    "\telse     : path = 'graph-occlusion-relevance-w{}-s{}-full.png'.format(OCCLUSION_SIZE, OCCLUSION_STRIDE)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, path),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01efc4c0-bb24-4be7-ac40-871510fa132b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Setting x ticks\n",
    "\n",
    "def set_xticks_zero_right (ticks, left, right) : return ticks - right\n",
    "def set_xticks_zero_left  (ticks, left, right) : return ticks - left"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81891ff4-6251-49b3-b6d7-05cc798a9a0c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot relevance for each region seperately\n",
    "\n",
    "lengths = [1000, 300, 350, 500]\n",
    "regions = [\n",
    "\t('Promoter',      0, sum(lengths[:1])),\n",
    "\t('5\\'UTR',     1000, sum(lengths[:2])),\n",
    "\t('3\\'UTR',     1300, sum(lengths[:3])),\n",
    "\t('Terminator', 1650, sum(lengths[:4])),\n",
    "]\n",
    "\n",
    "LINEWIDTH = 2\n",
    "ALPHA     = 0.8\n",
    "NROWS     = 2\n",
    "NCOLS     = 2\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 48})\n",
    "seaborn.set_theme(font_scale = 4)\n",
    "\n",
    "fig, ax = matplotlib.pyplot.subplots(nrows = NROWS, ncols = NCOLS, sharey = True, figsize = (NCOLS * 16, NROWS * 10))\n",
    "fig.tight_layout()\n",
    "\n",
    "for index, region in enumerate(regions) :\n",
    "\tif NROWS > 1 : axis = ax[index // NCOLS, index % NCOLS]\n",
    "\telse         : axis = ax[index]\n",
    "\n",
    "\tif   region[0] in ['Promoter']   : marker = 'TSS'\n",
    "\telif region[0] in ['5\\'UTR']     : marker = 'CDS'\n",
    "\telif region[0] in ['3\\'UTR']     : marker = 'TTS'\n",
    "\telif region[0] in ['Terminator'] : marker = 'TTS'\n",
    "\telse : raise ValueError()\n",
    "\n",
    "\tsubset = wocclusion_csv\n",
    "\tsubset = subset[subset['Midpoint'] >= region[1]]\n",
    "\tsubset = subset[subset['Midpoint'] <  region[2]]\n",
    "\n",
    "\tif region[0] in ['Promoter']   : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['5\\'UTR']     : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['3\\'UTR']     : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['Terminator'] : subset['Midpoint'] = set_xticks_zero_left (subset['Midpoint'], region[1], region[2])\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = subset,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Washburn',\n",
    "\t\tax        = axis,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tsubset = zocclusion_csv\n",
    "\tsubset = subset[subset['Midpoint'] >= region[1]]\n",
    "\tsubset = subset[subset['Midpoint'] <  region[2]]\n",
    "\n",
    "\tif region[0] in ['Promoter']   : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['5\\'UTR']     : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['3\\'UTR']     : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['Terminator'] : subset['Midpoint'] = set_xticks_zero_left (subset['Midpoint'], region[1], region[2])\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = subset,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Zrimec',\n",
    "\t\tax        = axis,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\taxis.set_title(region[0])\n",
    "\n",
    "\taxis.set_xlabel('Position from {}'.format(marker))\n",
    "\taxis.set_ylabel('Relevance')\n",
    "\n",
    "\taxis.legend(loc = 'upper center')\n",
    "\n",
    "matplotlib.pyplot.subplots_adjust(\n",
    "\tleft   = None,\n",
    "\tbottom = None,\n",
    "\tright  = None,\n",
    "\ttop    = None,\n",
    "\twspace = 0.05,\n",
    "\thspace = 0.40\n",
    ")\n",
    "\n",
    "matplotlib.pyplot.savefig(\n",
    "\tos.path.join(OUT_DATA, 'graph-occlusion-relevance-w{}-s{}-seperate.png'.format(OCCLUSION_SIZE, OCCLUSION_STRIDE)),\n",
    "\tdpi         = 120,\n",
    "\tformat      = 'png',\n",
    "\tbbox_inches = 'tight',\n",
    "\tpad_inches  = 0\n",
    ")\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e02c67c3-b738-43c7-8bfd-b703fb411008",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot relevance for entire sequence\n",
    "\n",
    "RELEVANCE_TYPE = 'R2'\n",
    "PLOT_TWINX     = False\n",
    "LINEWIDTH      = 2\n",
    "ALPHA          = 0.8\n",
    "\n",
    "for vline in [False, True] :\n",
    "\tfig, ax1 = matplotlib.pyplot.subplots(figsize = (16, 8))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = wocclusion_json,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Washburn',\n",
    "\t\tax        = ax1,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tif PLOT_TWINX : ax2 = ax1.twinx()\n",
    "\telse          : ax2 = ax1\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = zocclusion_json,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Zrimec',\n",
    "\t\tax        = ax2,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tax1.axhline(WASHBURN_BASELINE[RELEVANCE_TYPE], color = 'b', linestyle = '--', alpha = ALPHA)\n",
    "\tax2.axhline(ZRIMEC_BASELINE[RELEVANCE_TYPE],   color = 'r', linestyle = '--', alpha = ALPHA)\n",
    "\n",
    "\tif vline : \n",
    "\t\tax1.axvline(1000, ymin = 0, ymax = 1, color = 'k')\n",
    "\t\tax1.axvline(1300, ymin = 0, ymax = 1, color = 'k')\n",
    "\t\tax1.axvline(1650, ymin = 0, ymax = 1, color = 'k')\n",
    "\n",
    "\tax1.set_xlabel('Position')\n",
    "\n",
    "\tif PLOT_TWINX : ax2.set_ylabel('Zrimec '   + RELEVANCE_TYPE)\n",
    "\tif PLOT_TWINX : ax1.set_ylabel('Washburn ' + RELEVANCE_TYPE)\n",
    "\telse          : ax1.set_ylabel(RELEVANCE_TYPE)\n",
    "\n",
    "\tif vline : path = 'graph-occlusion-{}-w{}-s{}-full-with-vline.png'.format(RELEVANCE_TYPE.lower(), OCCLUSION_SIZE, OCCLUSION_STRIDE)\n",
    "\telse     : path = 'graph-occlusion-{}-w{}-s{}-full.png'.format(RELEVANCE_TYPE.lower(), OCCLUSION_SIZE, OCCLUSION_STRIDE)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, path),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1468d107-99da-43e5-ae89-50e9c99768a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot relevance for each region seperately\n",
    "\n",
    "lengths = [1000, 300, 350, 500]\n",
    "regions = [\n",
    "\t('Promoter',      0, sum(lengths[:1])),\n",
    "\t('5\\'UTR',     1000, sum(lengths[:2])),\n",
    "\t('3\\'UTR',     1300, sum(lengths[:3])),\n",
    "\t('Terminator', 1650, sum(lengths[:4])),\n",
    "]\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 48})\n",
    "seaborn.set_theme(font_scale = 4)\n",
    "\n",
    "LINEWIDTH = 2\n",
    "ALPHA     = 0.8\n",
    "NROWS     = 2\n",
    "NCOLS     = 2\n",
    "\n",
    "fig, ax = matplotlib.pyplot.subplots(nrows = NROWS, ncols = NCOLS, sharey = True, figsize = (NCOLS * 16, NROWS * 10))\n",
    "fig.tight_layout()\n",
    "\n",
    "for index, region in enumerate(regions) :\n",
    "\tif NROWS > 1 : axis = ax[index // NCOLS, index % NCOLS]\n",
    "\telse         : axis = ax[index]\n",
    "\n",
    "\tif   region[0] in ['Promoter']   : marker = 'TSS'\n",
    "\telif region[0] in ['5\\'UTR']     : marker = 'CDS'\n",
    "\telif region[0] in ['3\\'UTR']     : marker = 'TTS'\n",
    "\telif region[0] in ['Terminator'] : marker = 'TTS'\n",
    "\telse : raise ValueError()\n",
    "\n",
    "\tsubset = wocclusion_json\n",
    "\tsubset = subset[subset['Midpoint'] >= region[1]]\n",
    "\tsubset = subset[subset['Midpoint'] <  region[2]]\n",
    "\n",
    "\tif region[0] in ['Promoter']   : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['5\\'UTR']     : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['3\\'UTR']     : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['Terminator'] : subset['Midpoint'] = set_xticks_zero_left (subset['Midpoint'], region[1], region[2])\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = subset,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Washburn',\n",
    "\t\tax        = axis,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tsubset = zocclusion_json\n",
    "\tsubset = subset[subset['Midpoint'] >= region[1]]\n",
    "\tsubset = subset[subset['Midpoint'] <  region[2]]\n",
    "\n",
    "\tif region[0] in ['Promoter']   : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['5\\'UTR']     : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['3\\'UTR']     : subset['Midpoint'] = set_xticks_zero_right(subset['Midpoint'], region[1], region[2])\n",
    "\tif region[0] in ['Terminator'] : subset['Midpoint'] = set_xticks_zero_left (subset['Midpoint'], region[1], region[2])\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = subset,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Zrimec',\n",
    "\t\tax        = axis,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\taxis.set_title(region[0])\n",
    "\n",
    "\taxis.set_xlabel('Position from {}'.format(marker))\n",
    "\taxis.set_ylabel(RELEVANCE_TYPE)\n",
    "\n",
    "\taxis.axhline(WASHBURN_BASELINE[RELEVANCE_TYPE], color = 'b', linestyle = '--', alpha = ALPHA)\n",
    "\taxis.axhline(ZRIMEC_BASELINE[RELEVANCE_TYPE],   color = 'r', linestyle = '--', alpha = ALPHA)\n",
    "\n",
    "\taxis.legend(loc = 'center right')\n",
    "\n",
    "matplotlib.pyplot.subplots_adjust(\n",
    "\tleft   = None,\n",
    "\tbottom = None,\n",
    "\tright  = None,\n",
    "\ttop    = None,\n",
    "\twspace = 0.05,\n",
    "\thspace = 0.40\n",
    ")\n",
    "\n",
    "matplotlib.pyplot.savefig(\n",
    "\tos.path.join(OUT_DATA, 'graph-occlusion-{}-w{}-s{}-seperate.png'.format(RELEVANCE_TYPE.lower(), OCCLUSION_SIZE, OCCLUSION_STRIDE)),\n",
    "\tdpi         = 120,\n",
    "\tformat      = 'png',\n",
    "\tbbox_inches = 'tight',\n",
    "\tpad_inches  = 0\n",
    ")\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c07fca3e-19a1-4512-8474-a75469cb7252",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot relevance and r2 for entire sequence\n",
    "\n",
    "LINEWIDTH = 2\n",
    "ALPHA     = 0.8\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 36})\n",
    "seaborn.set_theme(font_scale = 3)\n",
    "\n",
    "for vline in [True, False] :\n",
    "\tfig, ax = matplotlib.pyplot.subplots(nrows = 2, ncols = 1, sharex = True, figsize = (16, 16))\n",
    "\tfig.tight_layout()\n",
    "\n",
    "\tax1 = ax[0]\n",
    "\tax2 = ax[1]\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = wocclusion_csv,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Washburn',\n",
    "\t\tax        = ax1,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = zocclusion_csv,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Zrimec',\n",
    "\t\tax        = ax1,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tif vline : \n",
    "\t\tax1.axvline(1000, ymin = 0, ymax = 1, color = 'k')\n",
    "\t\tax1.axvline(1300, ymin = 0, ymax = 1, color = 'k')\n",
    "\t\tax1.axvline(1650, ymin = 0, ymax = 1, color = 'k')\n",
    "\n",
    "\t\t# ax1.set_title('{:>25s} {:>36s} {:>15s} {:>26s}'.format('Promoter', '5\\'UTR', '3\\'UTR', 'Terminator'))\n",
    "\t\tax1.set_title('{:>18s} {:>22s} {:>7s} {:>16s}'.format('Promoter', '5\\'UTR', '3\\'UTR', 'Terminator'))\n",
    "\n",
    "\tax1.set_xlabel('Position')\n",
    "\tax1.set_ylabel('Relevance')\n",
    "\n",
    "\tax1.legend(loc = 'upper left')\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = wocclusion_json,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Washburn',\n",
    "\t\tax        = ax2,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tif PLOT_TWINX : ax3 = ax2.twinx()\n",
    "\telse          : ax3 = ax2\n",
    "\n",
    "\tseaborn.lineplot(\n",
    "\t\tdata      = zocclusion_json,\n",
    "\t\tx         = 'Midpoint',\n",
    "\t\ty         = RELEVANCE_TYPE,\n",
    "\t\tlinewidth = LINEWIDTH,\n",
    "\t\tlabel     = 'Zrimec',\n",
    "\t\tax        = ax3,\n",
    "\t\talpha     = ALPHA\n",
    "\t)\n",
    "\n",
    "\tax2.axhline(WASHBURN_BASELINE[RELEVANCE_TYPE], color = 'b', linestyle = '--', alpha = ALPHA)\n",
    "\tax3.axhline(ZRIMEC_BASELINE[RELEVANCE_TYPE],   color = 'r', linestyle = '--', alpha = ALPHA)\n",
    "\n",
    "\tif vline : \n",
    "\t\tax2.axvline(1000, ymin = 0, ymax = 1, color = 'k')\n",
    "\t\tax2.axvline(1300, ymin = 0, ymax = 1, color = 'k')\n",
    "\t\tax2.axvline(1650, ymin = 0, ymax = 1, color = 'k')\n",
    "\n",
    "\tax2.set_xlabel('Position')\n",
    "\n",
    "\tif PLOT_TWINX : ax3.set_ylabel('Zrimec '   + RELEVANCE_TYPE)\n",
    "\tif PLOT_TWINX : ax2.set_ylabel('Washburn ' + RELEVANCE_TYPE)\n",
    "\telse          : ax2.set_ylabel(RELEVANCE_TYPE)\n",
    "\n",
    "\tif vline : path = 'graph-occlusion-both-w{}-s{}-full-with-vline.png'.format(OCCLUSION_SIZE, OCCLUSION_STRIDE)\n",
    "\telse     : path = 'graph-occlusion-both-w{}-s{}-full.png'.format(OCCLUSION_SIZE, OCCLUSION_STRIDE)\n",
    "\n",
    "\tmatplotlib.pyplot.savefig(\n",
    "\t\tos.path.join(OUT_DATA, path),\n",
    "\t\tdpi         = 120,\n",
    "\t\tformat      = 'png',\n",
    "\t\tbbox_inches = 'tight',\n",
    "\t\tpad_inches  = 0\n",
    "\t)\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "897559c6-0f8e-4e23-a9fe-37268e8df009",
   "metadata": {},
   "source": [
    "# 8. Relevance of Region Occlusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b587a484-5c51-4993-93cc-824e8538b143",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load regions occlusion data\n",
    "\n",
    "wregions = loader.load_json(os.path.join(wocclusion_path, 'occlusion', 'table-occlusion-region.json'))\n",
    "zregions = loader.load_json(os.path.join(zocclusion_path, 'occlusion', 'table-occlusion-region.json'))\n",
    "\n",
    "wregions.append({\n",
    "\t'end'           : numpy.inf,\n",
    "\t'mae'           : WASHBURN_BASELINE['MAE'],\n",
    "\t'mae-relevance' : 0.0,\n",
    "\t'mid'           : numpy.inf,\n",
    "\t'mse'           : WASHBURN_BASELINE['MSE'],\n",
    "\t'mse-relevance' : 0.0,\n",
    "\t'r2'            : WASHBURN_BASELINE['R2'],\n",
    "\t'r2-relevance'  : 0.0,\n",
    "\t'region'        : 'None',\n",
    "\t'start'         : numpy.inf\n",
    "})\n",
    "\n",
    "zregions.append({\n",
    "\t'end'           : numpy.inf,\n",
    "\t'mae'           : ZRIMEC_BASELINE['MAE'],\n",
    "\t'mae-relevance' : 0.0,\n",
    "\t'mid'           : numpy.inf,\n",
    "\t'mse'           : ZRIMEC_BASELINE['MSE'],\n",
    "\t'mse-relevance' : 0.0,\n",
    "\t'r2'            : ZRIMEC_BASELINE['R2'],\n",
    "\t'r2-relevance'  : 0.0,\n",
    "\t'region'        : 'None',\n",
    "\t'start'         : numpy.inf\n",
    "})\n",
    "\n",
    "wregions = pandas.DataFrame.from_dict(wregions)\n",
    "zregions = pandas.DataFrame.from_dict(zregions)\n",
    "\n",
    "wregions['Delta'] = wregions['r2'] - WASHBURN_BASELINE['R2']\n",
    "zregions['Delta'] = zregions['r2'] - ZRIMEC_BASELINE['R2']\n",
    "\n",
    "regions = pandas.concat([wregions.assign(Model = 'Washburn'), zregions.assign(Model = 'Zrimec')])\n",
    "regions = regions.rename(columns = {\n",
    "\t'start'        : 'Start',\n",
    "\t'region'       : 'Occlusion',\n",
    "\t'r2'           : 'R2',\n",
    "\t'r2-relevance' : 'Relevance',\n",
    "})\n",
    "\n",
    "regions = regions.sort_values(['Start', 'Model'])\n",
    "regions = regions.set_index(['Occlusion', 'Model'])\n",
    "regions = regions[['R2', 'Delta', 'Relevance']]\n",
    "\n",
    "regions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5211576d-ec32-44f0-81ac-ee49dcbd9421",
   "metadata": {},
   "source": [
    "# 9. Updating Group Barplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b39d2260-ac64-4771-bcbc-4564c96f9693",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define path\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\output\\nbp00-analysis'\n",
    "\n",
    "tissue_metadata = os.path.join(folder_path,  'tissue-metadata.csv')\n",
    "tissue_metadata = pandas.read_csv(tissue_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "957f8ec0-ced2-4e09-8635-504be84f6f1d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot barplots\n",
    "\n",
    "group1   = 'Tissue'\n",
    "group2   = 'Age'\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 42})\n",
    "seaborn.set_theme(font_scale = 3.33)\n",
    "\n",
    "unique1 = tissue_metadata[group1].unique()\n",
    "unique2 = tissue_metadata[group2].unique()\n",
    "counts1 = [len(tissue_metadata.loc[tissue_metadata[group1] == name]) for name in unique1]\n",
    "counts2 = [len(tissue_metadata.loc[tissue_metadata[group2] == name]) for name in unique2]\n",
    "\n",
    "df1 = pandas.DataFrame.from_dict({\n",
    "\tgroup1  : unique1,\n",
    "\t'Count' : counts1\n",
    "}).sort_values('Count', ascending = False)\n",
    "\n",
    "df2 = pandas.DataFrame.from_dict({\n",
    "\tgroup2  : unique2,\n",
    "\t'Count' : counts2\n",
    "}).sort_values('Count', ascending = False)\n",
    "\n",
    "fig, axis = matplotlib.pyplot.subplots(ncols = 2, nrows = 1, sharey = True, figsize = (28, 11))\n",
    "fig.tight_layout()\n",
    "\n",
    "g1 = seaborn.barplot(\n",
    "\tdata   = df1,\n",
    "\tx      = group1,\n",
    "\ty      = 'Count',\n",
    "\thue    = group1,\n",
    "\twidth  = 0.8,\n",
    "\tax     = axis[0],\n",
    "\talpha  = 0.9,\n",
    "\tdodge  = False\n",
    ")\n",
    "\n",
    "g2 = seaborn.barplot(\n",
    "\tdata   = df2,\n",
    "\tx      = group2,\n",
    "\ty      = 'Count',\n",
    "\thue    = group2,\n",
    "\twidth  = 0.8,\n",
    "\tax     = axis[1],\n",
    "\talpha  = 0.9,\n",
    "\tdodge  = False\n",
    ")\n",
    "\n",
    "handles, labels = axis[0].get_legend_handles_labels()\n",
    "axis[0].legend(handles = handles[1:], labels = labels[1:])\n",
    "\n",
    "handles, labels = axis[1].get_legend_handles_labels()\n",
    "axis[1].legend(handles = handles[1:], labels = labels[1:])\n",
    "\n",
    "g2.set(ylabel = None)\n",
    "g1.set(xticklabels = [])\n",
    "g2.set(xticklabels = [])\n",
    "\n",
    "matplotlib.pyplot.subplots_adjust(\n",
    "\tleft   = None,\n",
    "\tbottom = None,\n",
    "\tright  = None,\n",
    "\ttop    = None,\n",
    "\twspace = 0.05,\n",
    "\thspace = None\n",
    ")\n",
    "\n",
    "matplotlib.pyplot.savefig(\n",
    "\tos.path.join(OUT_DATA, 'class-distribution-tissue-age.png'),\n",
    "\tdpi         = 120,\n",
    "\tformat      = 'png',\n",
    "\tbbox_inches = 'tight',\n",
    "\tpad_inches  = 0\n",
    ")\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ccdc193-cbfe-4c94-b719-51858d893d75",
   "metadata": {},
   "source": [
    "# 10. Updating Region Length Histograms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b020b93e-be45-4053-8723-8462cc4a8ff5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define path\n",
    "\n",
    "folder_path = r'C:\\Developer\\Workspace\\PyCharm\\Projects\\upolanc-thesis\\output\\nbp00-analysis'\n",
    "\n",
    "gene_annotation = os.path.join(folder_path, 'gene-annotation.csv')\n",
    "gene_annotation = pandas.read_csv(gene_annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1084e09c-e5ef-42ce-9ad3-e879cb9adbc3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Group regions\n",
    "\n",
    "regions = ['mRNA', 'UTR5', 'CDS', 'UTR3']\n",
    "data    = gene_annotation\n",
    "groupby = 'Transcript'\n",
    "\n",
    "if regions is None :\n",
    "\tregions = data['Type'].unique().tolist()\n",
    "\n",
    "entries = list()\n",
    "\n",
    "for group, dataframe in data.groupby(groupby) :\n",
    "\tf0 = dataframe['Seq'].iloc[0]\n",
    "\tf1 = dataframe['Strand'].iloc[0]\n",
    "\tf3 = dataframe['Gene'].iloc[0]\n",
    "\tf4 = dataframe['Transcript'].iloc[0]\n",
    "\n",
    "\tfor region in regions :\n",
    "\t\tf2 = region\n",
    "\t\tf5 = dataframe['Exon'].iloc[0]\n",
    "\t\tf6 = dataframe['Parent'].iloc[0]\n",
    "\n",
    "\t\tdata = dataframe.loc[dataframe['Type'].isin([region])]\n",
    "\n",
    "\t\tf7 = -1 if len(data) == 0 else data['Start'].min()\n",
    "\t\tf8 = -1 if len(data) == 0 else data['End'].max()\n",
    "\t\tf9 =  0 if len(data) == 0 else data['Length'].sum()\n",
    "\n",
    "\t\tentries.append([f0, f1, f2, f3, f4, f5, f6, f7, f8, f9, len(data)])\n",
    "\n",
    "dataframe = pandas.DataFrame(\n",
    "\tdata    = entries,\n",
    "\tcolumns = ['Seq', 'Strand', 'Type', 'Gene', 'Transcript', 'Exon', 'Parent', 'Start', 'End', 'Length', 'Regions']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84c3d680-c2e7-47b7-ad68-7dcc4518238a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot region length histogram\n",
    "\n",
    "vlines = {\n",
    "\t'mRNA' : 0,\n",
    "\t'UTR5' : 300,\n",
    "\t'CDS'  : 0,\n",
    "\t'UTR3' : 350\n",
    "}\n",
    "\n",
    "FONTSIZE  = 42\n",
    "LINEWIDTH = 4\n",
    "ALPHA     = 0.9\n",
    "KDE       = False\n",
    "BINS      = 60\n",
    "LOG       = True\n",
    "VLINE     = True\n",
    "\n",
    "dataframe_utr5 = dataframe[dataframe['Type'] == 'UTR5']\n",
    "dataframe_utr5 = dataframe_utr5[dataframe_utr5['Length'] > 0]\n",
    "\n",
    "dataframe_utr3 = dataframe[dataframe['Type'] == 'UTR3']\n",
    "dataframe_utr3 = dataframe_utr3[dataframe_utr3['Length'] > 0]\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 36})\n",
    "seaborn.set_theme(font_scale = 3)\n",
    "\n",
    "fig, axis = matplotlib.pyplot.subplots(ncols = 2, nrows = 1, sharey = True, figsize = (26, 10))\n",
    "fig.tight_layout()\n",
    "\n",
    "seaborn.histplot(\n",
    "\tx         = 'Length',\n",
    "\tdata      = dataframe_utr5,\n",
    "\tax        = axis[0],\n",
    "\tlog_scale = LOG,\n",
    "\talpha     = ALPHA,\n",
    "\tkde       = KDE,\n",
    "\tbins      = BINS\n",
    ")\n",
    "\n",
    "seaborn.histplot(\n",
    "\tx         = 'Length',\n",
    "\tdata      = dataframe_utr3,\n",
    "\tax        = axis[1],\n",
    "\tlog_scale = LOG,\n",
    "\talpha     = ALPHA,\n",
    "\tkde       = KDE,\n",
    "\tbins      = BINS\n",
    ")\n",
    "\n",
    "if VLINE :\n",
    "\tif vlines['UTR5'] > 0 :\n",
    "\t\tpercentile = scipy.stats.percentileofscore(dataframe_utr5['Length'], vlines['UTR5'], nan_policy = 'omit')\n",
    "\n",
    "\t\tx = 1.08 * vlines['UTR5']\n",
    "\t\ty = 0.90 * axis[0].get_ylim()[-1]\n",
    "\n",
    "\t\taxis[0].axvline(vlines['UTR5'], color = '#F0665E', linewidth = LINEWIDTH, alpha = ALPHA)\n",
    "\t\taxis[0].text(x, y, f'{percentile:.1f}%', color = '#F0665E', fontsize = FONTSIZE, alpha = ALPHA)\n",
    "\n",
    "\tif vlines['UTR3'] > 0 :\n",
    "\t\tpercentile = scipy.stats.percentileofscore(dataframe_utr3['Length'], vlines['UTR3'], nan_policy = 'omit')\n",
    "\n",
    "\t\tx = 1.08 * vlines['UTR3']\n",
    "\t\ty = 0.90 * axis[1].get_ylim()[-1]\n",
    "\n",
    "\t\taxis[1].axvline(vlines['UTR3'], color = '#F0665E', linewidth = LINEWIDTH, alpha = ALPHA)\n",
    "\t\taxis[1].text(x, y, f'{percentile:.1f}%', color = '#F0665E', fontsize = FONTSIZE, alpha = ALPHA)\n",
    "\n",
    "axis[1].set_title('3\\'UTR')\n",
    "axis[0].set_title('5\\'UTR')\n",
    "axis[1].set_xlabel('Length')\n",
    "axis[0].set_xlabel('Length')\n",
    "axis[0].set_ylabel('Count')\n",
    "\n",
    "matplotlib.pyplot.subplots_adjust(\n",
    "\tleft   = None,\n",
    "\tbottom = None,\n",
    "\tright  = None,\n",
    "\ttop    = None,\n",
    "\twspace = 0.05,\n",
    "\thspace = None\n",
    ")\n",
    "\n",
    "matplotlib.pyplot.savefig(\n",
    "\tos.path.join(OUT_DATA, 'length-distribution.png'),\n",
    "\tdpi         = 120,\n",
    "\tformat      = 'png',\n",
    "\tbbox_inches = 'tight',\n",
    "\tpad_inches  = 0\n",
    ")\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca092c6a-5b81-4e4b-94e8-9ec529729b22",
   "metadata": {},
   "source": [
    "# 11. Tissue-Specific TPM Distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2ad77d8-0187-497d-8b39-fb83dc4b99c7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#\n",
    "\n",
    "anndata = loader.load_h5ad(filename = os.path.join(RES_NBP04, 'filter{}'.format(2), 'arabidopsis-r36.h5ad'))\n",
    "split   = loader.load_json(filename = os.path.join(RES_NBP01, 'filter{}'.format(2), 'filter.json'))\n",
    "\n",
    "keep_transcript = split['data']['keep_transcript']\n",
    "drop_transcript = split['data']['drop_transcript']\n",
    "\n",
    "keep = anndata.var.index.isin(keep_transcript)\n",
    "drop = anndata.var.index.isin(drop_transcript)\n",
    "\n",
    "keep = anndata[:, keep]\n",
    "drop = anndata[:, drop]\n",
    "\n",
    "tissues = keep.obs['Tissue'].value_counts().to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c50c59a9-3357-4c74-bf8f-9b5fcbfb9973",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#\n",
    "\n",
    "NROWS = 3\n",
    "NCOLS = 2\n",
    "ALPHA = 0.8\n",
    "\n",
    "matplotlib.rcParams.update({'font.size' : 48})\n",
    "seaborn.set_theme(font_scale = 4)\n",
    "\n",
    "fig, ax = matplotlib.pyplot.subplots(\n",
    "\tnrows   = NROWS,\n",
    "\tncols   = NCOLS,\n",
    "\tsharex  = True,\n",
    "\tsharey  = True,\n",
    "\tfigsize = (16 * NCOLS, 10 * NROWS)\n",
    ")\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "for index, (k, n) in enumerate(tissues.items()) :\n",
    "\trow = index // NCOLS\n",
    "\tcol = index  % NCOLS\n",
    "\n",
    "\taxis = ax[row, col]\n",
    "\n",
    "\tkeep_k = keep[keep.obs['Tissue'] == k, :].layers['boxcox1p'].mean(axis = 0)\n",
    "\tdrop_k = drop[drop.obs['Tissue'] == k, :].layers['boxcox1p'].mean(axis = 0)\n",
    "\n",
    "\tkeep_l = numpy.array(['Keep' for _ in keep_k])\n",
    "\tdrop_l = numpy.array(['Drop' for _ in drop_k])\n",
    "\n",
    "\tdf = pandas.DataFrame.from_dict({\n",
    "\t\t'TPM'   : numpy.concatenate((keep_k, drop_k)),\n",
    "\t\t'Label' : numpy.concatenate((keep_l, drop_l))\n",
    "\t})\n",
    "\n",
    "\tseaborn.histplot(\n",
    "\t\tdata  = df,\n",
    "\t\tx     = 'TPM',\n",
    "\t\tax    = axis,\n",
    "\t\thue   = 'Label',\n",
    "\t\talpha = ALPHA\n",
    "\t)\n",
    "\n",
    "\taxis.set_title(k.title())\n",
    "\n",
    "ax[2, 1].axis('off')\n",
    "\n",
    "matplotlib.pyplot.subplots_adjust(\n",
    "\tleft   = None,\n",
    "\tbottom = None,\n",
    "\tright  = None,\n",
    "\ttop    = None,\n",
    "\twspace = 0.05,\n",
    "\thspace = None\n",
    ")\n",
    "\n",
    "matplotlib.pyplot.savefig(\n",
    "\tos.path.join(OUT_DATA, 'tpm-distribution.png'),\n",
    "\tdpi         = 120,\n",
    "\tformat      = 'png',\n",
    "\tbbox_inches = 'tight',\n",
    "\tpad_inches  = 0\n",
    ")\n",
    "\n",
    "runtime.set_plot_theme()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
